### ISSUE TITLE
VictoriaLogs `stats` query pipe: `uniq_values()` processor consistently returning unrelated values, `values()` processor sometimes returning garbage data

### ISSUE THREAD

**Issue by Fusl** [VictoriaMetrics/VictoriaMetrics#7458]
## `uniq_values` processor

After trying to implement the workaround described in #7407, I noticed that the `uniq_values()` processor in the `stats` pipe consistently includes unrelated, non-matching values for entries that aren't even necessarily in the selected time-window. `values` doesn't seem to be affected by this but has a different issue (see below).

For example, the following query:

```
ClientIP:="ORIGINAL_IP" | stats count_uniq(ClientIP) count, uniq_values(ClientIP) ips
```

Sometimes returns the following result:

```json
[
  {
    "count": "1",
    "ips": "[\"UNRELATED_IPV6_1\",\"UNRELATED_IPV6_2\",\"UNRELATED_IPV4_1\",\"UNRELATED_IPV4_2\",\"ORIGINAL_IP\",\"UNRELATED_IPV4_3\",\"UNRELATED_IPV4_4\"]"
  }
]
```

At the same time, the following query:

```
ClientIP:="ORIGINAL_IP" | stats by (ClientIP) count_uniq(ClientIP) count, uniq_values(ClientIP) ips | unroll by (ips)
```

Ends up with the exact same unrelated values:

```
[
  {
    "ClientIP": "ORIGINAL_IP",
    "count": "1",
    "ips": "UNRELATED_IPV6_1"
  },
  {
    "ClientIP": "ORIGINAL_IP",
    "count": "1",
    "ips": "UNRELATED_IPV6_2"
  },
  {
    "ClientIP": "ORIGINAL_IP",
    "count": "1",
    "ips": "UNRELATED_IPV4_1"
  },
  {
    "ClientIP": "ORIGINAL_IP",
    "count": "1",
    "ips": "UNRELATED_IPV4_2"
  },
  {
    "ClientIP": "ORIGINAL_IP",
    "count": "1",
    "ips": "ORIGINAL_IP"
  },
  {
    "ClientIP": "ORIGINAL_IP",
    "count": "1",
    "ips": "UNRELATED_IPV4_3"
  },
  {
    "ClientIP": "ORIGINAL_IP",
    "count": "1",
    "ips": "UNRELATED_IPV4_4"
  }
]
```

As you can see, in both cases, the original value that was queried for is contained within the result set but other, unrelated values are includes as well, even though the result of `count_uniq(ClientIP)` is `1`.

---
## `values` processor

After debugging the above issue with the `uniq_values` processor, I wanted to ensure that the `values` processor does not produce the same result, but then noticed that, while it doesn't seem to be consistently returning the same, unrelated values, it does sometimes return garbage data:

```
RuleID:="<redacted>" | stats by (ClientASN,ClientASNDescription) count_uniq(ClientIP) count_uniq, count(ClientIP) count, values(ClientIP) ips | sort by (count_uniq) desc
```

Produces the following kinds of values on the first query:

```json
  {
    "count_uniq": "2",
    "ClientASN": "36231",
    "ClientASNDescription": "TEMPEST-HOSTING",
    "count": "11",
    "ips": "[\"<ip address>\",\"<ip address>\",\"6�\\u0000\\u0000��140613623\",\"6�\\u0000\\u0000��140613623\",\"<ip address>\",\"<ip address>\",\"ASN21859ClientA\",\"ASN21859ClientA\",\"ASN21859ClientA\",\"ASN21859ClientA\",\"ASN21859ClientA\"]"
  },
```
```json
  {
    "count_uniq": "3",
    "ClientASN": "60068",
    "ClientASNDescription": "CDN77 _",
    "count": "6",
    "ips": "[\"<ip address>\",\"<ip address>\",\"ientASN39351ClientASNDescriptionESAB\",\"<ip address>\",\"<ip address>\",\"<ip address>\"]"
  },
```
```json
  {
    "count_uniq": "4",
    "ClientASN": "12876",
    "ClientASNDescription": "Online SAS",
    "count": "7",
    "ips": "[\"<ip address>\",\"<ip address>\",\"<ip address>\",\"<ip address>\",\"<ip address>\",\"ientASN21859Cl\",\"<ip address>\"]"
  },
```
```json
  {
    "count_uniq": "5",
    "ClientASN": "25820",
    "ClientASNDescription": "IT7NET",
    "count": "8",
    "ips": "[\"<ip address>\",\"<ip address>\",\"<ip address>\",\"<ip address>\",\"\\u0001\\u0002�\\u000e` +��*+��\",\"<ip address>\",\"<ip address>\",\"LARENET2a09:ba\"]"
  },
```

All the values with `<ip address>` have been manually redacted, they contain valid, human-readable IP addresses.

Refreshing the query with the same time frame produces slightly different garbage data:

```json
  {
    "count_uniq": "2",
    "ClientASN": "36231",
    "ClientASNDescription": "TEMPEST-HOSTING",
    "count": "11",
    "ips": "[\"<ip address>\",\"6�\\u0000\\u0000��140613623\",\"6�\\u0000\\u0000��140613623\",\"<ip address>\",\"<ip address>\",\"` +��*+��\\u0001128.1\",\"` +��*+��\\u0001128.1\",\"` +��*+��\\u0001128.1\",\"` +��*+��\\u0001128.1\",\"` +��*+��\\u0001128.1\",\"<ip address>\"]"
  },
```

Even limiting the query with a `| ClientASN:="36231"` produces similar results for the same ASN, again different garbage data:

```json
[
  {
    "count_uniq": "2",
    "ClientASN": "36231",
    "ClientASNDescription": "TEMPEST-HOSTING",
    "count": "11",
    "ips": "[\"<ip address>\",\"Y-USER-CONTENT3\",\"Y-USER-CONTENT3\",\"Y-USER-CONTENT3\",\"Y-USER-CONTENT3\",\"Y-USER-CONTENT3\",\"<ip address>\",\"<ip address>\",\"<ip address>\",\".149.89233\",\"<ip address>\"]"
  }
]
```

### To Reproduce

1. Start up a new VictoriaLogs instance without any custom configuration
2. Feed with random VALID IPs and using separate streams:
```bash
while true; do
    ip1=$(($RANDOM%256))
    ip2=$(($RANDOM%256))
    ip3=$(($RANDOM%256))
    ip4=$(($RANDOM%256))
    ipsum=$(($ip1+$ip2+$ip3+$ip4))
    echo '{"_msg":"foo '$ipsum'","ClientIP":"'$ip1.$ip2.$ip3.$ip4'","ipsum":"'$ipsum'"}'
done \
| tee /dev/stderr \
| curl -s -XPOST -T /dev/stdin "http://127.0.0.1:9428/insert/jsonline?_stream_fields=ipsum
```
3. In another window, repeatedly query VictoriaLogs with `query=* | stats by(sum) count(ClientIP) count, uniq_values(ClientIP) uniq_values, values(ClientIP) values` to find INVALID IPs being returned in the `values` field
```bash
curl -s http://localhost:9428/select/logsql/query -d 'query=* | stats by(sum) count(ClientIP) count, uniq_values(ClientIP) uniq_values, values(ClientIP) values' \
| jq -rs '. | map(.values = (.values | fromjson)) | map(.values)[][]' \
| grep -vE '^([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])$'
```
4. The query may take a long time to execute but after a couple times, it'll return invalid IPs such as `1825.103.76.2`, `����31.1`, `12.101.395`, `2.246.121178`, `S▒<A-F&a6`, `164.4082.237.`, ...

### Version

victoria-logs-20241030-223856-tags-v0.40.0-victorialogs-0-g1a7b55009b

**Comment by Fusl** [comment]
Updated with reproduction steps

**Comment by valyala** [comment]
@Fusl , thanks for the detailed bugreport! It has been appeared that you caught two distinct bugs:

- The bug in the [`values` stats function](https://docs.victoriametrics.com/victorialogs/logsql/#values-stats), which may result in garbage output. This bug has been fixed in the commit 2f1ce74d97184215a20a83e5cf4ed2a1205c37c6 .
- The bug in the [`uniq_values` stats function](https://docs.victoriametrics.com/victorialogs/logsql/#uniq_values-stats), which may result in values, which do not match the specified [query filters](https://docs.victoriametrics.com/victorialogs/logsql/#filters). This bug has been fixed in the commit 546bf7d579d0c19134b8a51835edc540fd00c943 .

These commits will be included in the next release. In the mean time you can build VictoriaLogs from the commit 546bf7d579d0c19134b8a51835edc540fd00c943 according to [these instructions](https://docs.victoriametrics.com/victorialogs/quickstart/#building-from-source-code) and verify whether both bugfixes work as expected.

**Comment by Fusl** [comment]
Tested `victoria-logs-20241108-224910-heads-master-0-g23aa1897c` with the same queries that previously returned garbage or unrelated data and seems to be working fine now. Thank you for the fix.

**Comment by valyala** [comment]
@Fusl , thanks for the confirmation that bugs are fixed now!

Both bugfixes have been included in [v0.42.0-victorialogs](https://docs.victoriametrics.com/victorialogs/changelog/). Closing the issue as fixed then.
